File: src/controller/java/src/chip/devicecontroller/OperationalKeyConfig.java
package chip.devicecontroller;

import androidx.annotation.Nullable;

/** Represents a set of operating credentials for a ChipDeviceController. */
public final class OperationalKeyConfig {
  private final byte[] trustedRootCertificate;
  @Nullable private final byte[] intermediateCertificate;
  private final byte[] nodeOperationalCertificate;
  private final byte[] ipkEpochKey;
  private final KeypairDelegate keypairDelegate;

  /**
   * @param keypairDelegate a delegate for signing operations
   * @param trustedRootCertificate the trusted root X.509 certificate in DER-encoded form
   * @param intermediateCertificate the intermediate X.509 certificate in DER-encoded form
   * @param nodeOperationalCertificate the node operational X.509 certificate in DER-encoded form
   * @param ipkEpochKey the IPK epoch key to use (assuming a single one)
   */
  public OperationalKeyConfig(
      KeypairDelegate keypairDelegate,
      byte[] trustedRootCertificate,
      @Nullable byte[] intermediateCertificate,
      byte[] nodeOperationalCertificate,
      byte[] ipkEpochKey) {
    this.keypairDelegate = keypairDelegate;
    this.trustedRootCertificate = trustedRootCertificate.clone();
    this.intermediateCertificate =
        intermediateCertificate == null ? null : intermediateCertificate.clone();
    this.nodeOperationalCertificate = nodeOperationalCertificate.clone();
    this.ipkEpochKey = ipkEpochKey.clone();
  }

  /** Returns the trusted root X.509 certificate in DER-encoded form. */
  public byte[] getTrustedRootCertificate() {
    return trustedRootCertificate.clone();
  }

  /** Returns the intermediate X.509 certificate in DER-encoded form, if provided. */
  @Nullable
  public byte[] getIntermediateCertificate() {
    return intermediateCertificate == null ? null : intermediateCertificate.clone();
  }

  /** Returns the node operational X.509 certificate in DER-encoded form. */
  public byte[] getNodeOperationalCertificate() {
    return nodeOperationalCertificate.clone();
  }

  /** Returns the IPK epoch key. */
  public byte[] getIpkEpochKey() {
    return ipkEpochKey.clone();
  }

  /** Returns the delegate for signing operations. */
  public KeypairDelegate getKeypairDelegate() {
    return keypairDelegate;
  }
}

File: src/controller/java/src/matter/controller/OperationalKeyConfig.kt
/*
 *   Copyright (c) 2023 Project CHIP Authors
 *   All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
package matter.controller

import chip.devicecontroller.KeypairDelegate

/**
 * CertificateData Configuration data for X.509 certificates.
 *
 * @property trustedRootCertificate The trusted root X.509 certificate in DER-encoded form.
 * @property intermediateCertificate The optional intermediate X.509 certificate in DER-encoded
 *   form.
 * @property operationalCertificate The node operational X.509 certificate in DER-encoded form.
 */
class CertificateData(
  val trustedRootCertificate: ByteArray,
  val intermediateCertificate: ByteArray?,
  val operationalCertificate: ByteArray
)

/**
 * Configuration for use with CASE (Chip Authentication Session Establishment) session
 * establishment.
 *
 * @property keypairDelegate A delegate for signing operations.
 * @property certificateData Configuration data for X.509 certificates.
 * @property ipk The Identity Protection Key.
 * @property fabricId The fabric ID to which these operational credentials are associated.
 * @property nodeId The Admin Node ID to which these operational credentials are associated.
 */
class OperationalKeyConfig(
  val keypairDelegate: KeypairDelegate,
  val certificateData: CertificateData,
  val ipk: ByteArray,
  val fabricId: Long,
  val nodeId: Long
)
